modules/biobank/php/specimencontroller.class.inc

```diff
@@ -0,0 +1,531 @@
+ * Biobank Module.
+ *
+ * PHP Version 7.2
+ *
+ * @category   Controller
+ * @package    Loris
+ * @subpackage Biobank
+ * @author     Henri Rabalais <henri.rabalais@mcin.ca>
+ * @license    http://www.gnu.org/licenses/gpl-3.0.text GPLv3
+ * @link       http://www.github.com/aces/Lors/
+ */
+namespace LORIS\biobank;
+
+/**
+ * This class features the code that enables access to the Specimens of the LORIS
+ * Biobank Module.
+ *
+ * PHP Version 7.2
+ *
+ * @category   Controller
+ * @package    Loris
+ * @subpackage Biobank
+ * @author     Henri Rabalais <henri.rabalais@mcin.ca>
+ * @license    http://www.gnu.org/licenses/gpl-3.0.text GPLv3
+ * @link       http://www.github.com/aces/Lors/
+ */
+class SpecimenController
+{
+    /**
+     * Private instances of the Database, User and ContainerDAO classes
+     *
+     * @var \Database   $db
+     * @var \User       $user
+     * @var SpecimenDAO $dao
+     */
+    private $db, $user, $dao;
+
+    /**
+     * Specimen Controller Class constructor
+     *
+     * @param Database $db   Database Instance.
+     * @param User     $user User Instance.
+     */
+    function __construct(\Database $db, \User $user)
+    {
+        $this->db   = $db;
+        $this->user = $user;
+        $this->dao  = $this->getDataProvisioner();
+    }
+
+    /**
+     * Get all Specimen Objects permitted by the Specimen Data Provisioner (DAO).
+     *
+     * @return array $specimens All permissable Specimen Objects.
+     */
+    public function getInstances() : array
+    {
+        $this->validatePermission('GET');
+        $specimens  = [];
+        $specimenIt = $this->dao->execute($this->user);
+        foreach ($specimenIt as $id => $specimen) {
+            $specimens[$id] = $specimen;
+        }
+
+        return $specimens;
+    }
+
+    /**
+     * This method takes in a list of specimen and container arrays from the
+     * front-end. The specimens are validated while the containers are
+     * sent to their respective controller to be validated and saved. The ids of
+     * the saved containers are set to the specimen instances which are then
+     * sent to the DAO to be saved.
+     *
+     * @param array $list a list of specimen and container arrays to be created.
+     *
+     * @return void
+     */
+    public function createInstances(array $list) : void
+    {
+        $this->validatePermission('POST');
+
+        foreach ($list as $i => $item) {
+            $containerList[$i] = $item['container'];
+            $specimenList[$i]  = $this->prepare($item['specimen']);
+        }
+
+        $contController = new ContainerController($this->db, $this->user);
+        $containerIds   = $contController->createInstances($containerList);
+
+        foreach ($specimenlist as $i => $specimen) {
+            $specimen->setContainerId($containerIds[$i]);
+            $this->dao->saveInstance($specimen);
+        }
+    }
+
+    // TODO: Decide if this function should take a list of specimens, rather
```

Comment:
Yeah, the only issue is that when many entities are sent to the back from updating simultaneously, such as we several containers are checked out of a parent container at the same time. This can be resolved through a forloop on the client-side. This does mean though that some checkouts may succeed while others fail. 

That's the only realy use case I can think of at the moment.

Reply:
hmmm... i see. the question is then, do we fail all updates if one of them fails...

let's discuss that at somepoint